### ISSUE TITLE
ES.107: Signed subscripts for containers?!

### ISSUE THREAD

**Issue by imre-palik** [isocpp/CppCoreGuidelines#1229]
If the "proper" way would be signed subscripts, then could somebody explain me what would be the semantic meaning of a negative subscript e.g., for a std::vector?

BTW, I would find it a really bad surprise if the following code would fail for any container

`for (auto i = vec.size() -1; i  < vec.size(); i -= 2)
    cout << vec[i] << '\n'`

**Comment by jwakely** [comment]
Ah, the old "start big and keep getting smaller until you get enormous" style of counting understood by school children and beginners everywhere. /sarcasm

The point of using signed numbers is not to allow negative sizes or negative indices, it's to avoid the modulus arithmetic (i.e. wrapping) behaviour of unsigned integers in C++. Because that's not how counting works.

**Comment by imre-palik** [comment]
So defining interfaces that are semantically bad, and moving to undefined behaviour land with under/overflow + negative values is so much better, than well defined modulus arithmetic? /sarcasm

BTW, my daughter just learnt this "start big and keep getting smaller until you get enormous" style of counting in school.  She is in 2nd grade.  (OK, with enormous <= 100)

**Comment by MikeGitb** [comment]
@imre-palik re-palik: Do your really think 

> for (auto i = vec.size() -1; i < vec.size(); i -= 2) cout << vec[i] << '\n';

is more intuitive than

    for (auto i = vec.size() -1; i >= 0; i -= 2) cout << vec[i] << '\n';

The main (and imho only) problem with signed indices is inconsistency with the standard library :(

**Comment by Brown-LiuZan** [comment]
It seems the second one is an infinite loop.



Best regards and thanks
Brown.LiuZan
Smartphone: +86 18600694846<tel:+8618600694846>
在 MikeGitb <notifications@github.com>，2018年6月12日 上午12:19写道：

@imre-palik<https://github.com/imre-palik> re-palik: Do your really think

for (auto i = vec.size() -1; i < vec.size(); i -= 2) cout << vec[i] << '\n';

is more intuitive than

for (auto i = vec.size() -1; i >= 0; i -= 2) cout << vec[i] << '\n';


The main (and imho only) problem with signed indices is inconsistency with the standard library :(

**Comment by jwakely** [comment]
It's an infinite loop only because the type is unsigned. It's obviously more natural to count from the end down to the beginning, not from the end down to above the end. But the obvious and natural loop doesn't work because of unsigned behaviour.

**Comment by MikeGitb** [comment]
@Brown-LiuZan: The question was about signed subscripts for containers. I just wanted to demonstrate how much (imho) more intuitive you coul write such a loop if the indices and size were signed. But maybe I should not have used auto here to make this more clear.

**Comment by imre-palik** [comment]
Signed indices mean undefined behaviour on the interface.  I think it is a bad idea.  Consider the following loop:

`for (decltype(vec.size) i = -1; i >= -1 * vec.size(); i -= 2) cout << vec[i] << endl;`

What would this mean?

Keep also in mind that the following can be well defined for properly chosen l, if foo is a pointer:

`for (int i = -1; i > l; i -= 2) cout << foo[i] << endl;`

**Comment by jwakely** [comment]
What's your point? That it's possible to write confusing code in C++?

Nobody is suggesting that using negative indices with `std::vector` makes sense. You're attacking a strawman.

**Comment by jwakely** [comment]
Also, newsflash: containers already have undefined behaviour in their interfaces.

**Comment by MikeGitb** [comment]
Why on Earth would you write such code? This really gets rediculous. Shall I now answer by putting together some meaningless code that exercises UB with vector's current interface?

Passing a negative integer as an index is precisely as sensible as passing a positive value that is too big. Both are UB, both can be checked in debug mode internally.

BTW. The equivalent for your suitable chosen pointer `foo` example would be if `vec` would be a suitable chosen vector::iterator not a vector and the equivalent to the vector would be if `f` would be an array or a pointer to the start of the array, in which case your second example would be just as meaningless/UB

**Comment by jwakely** [comment]
> If I see that a function accepts a signed parameter, then I assume that negative values are meaningful there. Otherwise, that parameter would be unsigned. (Self documenting code.)

If I see a function that accepts an unsigned parameter, then I assume that modular arithmetic is desired there. Otherwise, that parameter would be signed. (Self documenting code).

Gee, it's almost like unsigned integers have two properties, and sometimes you only want one of them. And the choice of when to use them or not can be guided by useful conventions, but not by some absolute truth.

> My point here is that allowing negative values for indices is possible, but it won't be consistent with builtin types.

I'll repeat, "The point of using signed numbers is not to allow negative sizes or negative indices". This discussion seems pointless when you ignore what anybody else says.

**Comment by imre-palik** [comment]
I didn't ignore it, but maybe I was not clear enough with my response.  (BTW, ad hominem is neither nice, nor polite)  So here is my position again:

I don't think it is a good tradeoff to exchange something well and widely understood, like modular arithmetic to something weird, like possibility of negative indices.  Especially not if the negative values are simply undefined.

For the backward looping you will either need modular arithmetic and unsigned indices, or non modular arithmetic and signed indices.  So if you have unsigned indices, then modular arithmetic is desired there.  So, at least in this case, you want both properties.

**Comment by MikeGitb** [comment]
@imre-palik: If I write code, I write it with a purpose, i.e. I write it because I want it to achieve something not just because I can. I e.g. don't litter my code with arbitrary no-ops like `{int i=0;}` just because I can. Now I'm asking you: why would you write such code? What would you expect that code to achieve? Why would you use a negative index on a vector if you don't know what it means? Wouldn't you consult the documentation before doing that? 

I understand that people want to prevent code where the programmer expected it to do `A` but in reallity it really does `B`, but here I don't see, what `A` (the intended purpose) in your example is.

> If I see that a function accepts a signed parameter, then I assume that negative values are meaningful there.

Narrow contracts are pretty common in c++ apis. `std::vector::operator[](size_t i)` already has a narrow contract even with unsigned indexces (i.e. `i` must be smaller than `size()`). If you are in the habbit of passing any possible value of a parameter type to a function just because you can (not because you expect it to do something meaningful), c++ is not the right language for you.

Btw.: The recently merged proposal for contracts would certainly help with the "self-documenting" aspect

Anyway, I think I have sufficiently expressed my opinion about your arguments and the reason for it. Unless you come up with new arguments (and I would encourage you to first have a look at past discussions about signed / unsigned index types) I'm out.

**Comment by imre-palik** [comment]
Yes, more words are spent on this topic than it worth.  Oddly I don't get that much comments on those issues I find way more interesting ...

Anyway, if you think of std::vector::operator[]() as a performance optimised version of std::vector::at(), then you can argue for the need of the narrowing contract.  I would loved to see some reasoning along these lines here for the signed indices.  (Then I would surely learnt something.)
